### PR TITLE
[ticket/14990] Add core events to the Twig environment

### PR DESCRIPTION
**Pull request by rxu**
PHPBB3-14990

Checklist:

- [x] Correct branch: master for new features; 3.2.x, 3.1.x for fixes
- [x] Tests pass
- [x] Code follows coding guidelines: [master / 3.2.x](https://area51.phpbb.com/docs/master/coding-guidelines.html), [3.1.x](https://area51.phpbb.com/docs/31x/coding-guidelines.html)
- [x] Commit follows commit message [format](https://wiki.phpbb.com/Git#Commit_Messages)

Tracker ticket (set the ticket ID to **your ticket ID**):

https://tracker.phpbb.com/browse/PHPBB3-14990

### CODE REVIEW

**Comment by rxu** [comment]
!set Event

**Comment by DavidIQ** [comment]
Aren't these related to what this event would do or adding some similar ability without the use of events?
https://github.com/phpbb/phpbb/pull/4605
https://github.com/phpbb/phpbb/pull/4612

**Comment by rxu** [comment]
Not really, as far as I see those 2 PRs are related to the template vars and can;'t allow changing html output.

**Comment by javiexin** [comment]
I agree with rxu, these are complementary, but not quite the same thing.
The two mentioned PRs, plus https://github.com/phpbb/phpbb/pull/4598, are all manipulating template blocks (variables), nothing to do with the rendering.  There are potentially some overlapping use cases, but there are definitely cases to use both in very different ways.

**Comment by Nicofuma** [comment]
I don't really like this, but I agree that we don't really have the choice right now :/

**Comment by rxu** [comment]
Agreed, but probably better than nothing.

File: phpBB/phpbb/template/twig/environment.php [hunk 1]
```diff
@@ -54,14 +57,16 @@ class environment extends \Twig_Environment
 	* @param \phpbb\extension\manager $extension_manager phpBB extension manager
 	* @param \Twig_LoaderInterface $loader Twig loader interface
 	* @param array $options Array of options to pass to Twig
+	* @param \phpbb\event\dispatcher_interface	$phpbb_dispatcher	Event dispatcher object
 	*/
-	public function __construct(\phpbb\config\config $phpbb_config, \phpbb\filesystem\filesystem $filesystem, \phpbb\path_helper $path_helper, $cache_path, \phpbb\extension\manager $extension_manager = null, \Twig_LoaderInterface $loader = null, $options = array())
+	public function __construct(\phpbb\config\config $phpbb_config, \phpbb\filesystem\filesystem $filesystem, \phpbb\path_helper $path_helper, $cache_path, \phpbb\extension\manager $extension_manager = null, \Twig_LoaderInterface $loader = null, $options = array(), \phpbb\event\dispatcher_interface	$phpbb_dispatcher = null)
```
Comment (Elsensee) [inline phpBB/phpbb/template/twig/environment.php]: This tab is a bit weird to me.

**Comment by Elsensee** [comment]
What do you want to do with this event what you can't do with the already mentioned events?

**Comment by rxu** [comment]
As I said above 😉 
> those 2 PRs are related to the template vars and **don't allow changing html output**.

File: phpBB/phpbb/template/twig/environment.php (hunk 1 above)
Comment (rxu) [inline phpBB/phpbb/template/twig/environment.php]: Tab?

Comment (Elsensee) [inline phpBB/phpbb/template/twig/environment.php]: Yeah, between `\phpbb\event\dispatcher_interface` and `$phpbb_dispatcher` isn't a normal space but a tab.

Comment (rxu) [inline phpBB/phpbb/template/twig/environment.php]: Oh, indeed. Good catch 😃

**Comment by rxu** [comment]
Sure, thanks.

**Comment by rxu** [comment]
Tests failure is unrelated.

File: phpBB/phpbb/template/twig/environment.php [hunk 2]
```diff
@@ -54,14 +57,16 @@ class environment extends \Twig_Environment
 	* @param \phpbb\extension\manager $extension_manager phpBB extension manager
 	* @param \Twig_LoaderInterface $loader Twig loader interface
 	* @param array $options Array of options to pass to Twig
+	* @param \phpbb\event\dispatcher_interface	$phpbb_dispatcher	Event dispatcher object
 	*/
-	public function __construct(\phpbb\config\config $phpbb_config, \phpbb\filesystem\filesystem $filesystem, \phpbb\path_helper $path_helper, $cache_path, \phpbb\extension\manager $extension_manager = null, \Twig_LoaderInterface $loader = null, $options = array())
+	public function __construct(\phpbb\config\config $phpbb_config, \phpbb\filesystem\filesystem $filesystem, \phpbb\path_helper $path_helper, $cache_path, \phpbb\extension\manager $extension_manager = null, \Twig_LoaderInterface $loader = null, $options = array(), \phpbb\event\dispatcher_interface $phpbb_dispatcher = null)
```
Comment (marc1706) [inline phpBB/phpbb/template/twig/environment.php]: Objects being passd to classes should always come before any data input like arrays or strings. Please move the dispatcher in front of the options array.

Comment (rxu) [inline phpBB/phpbb/template/twig/environment.php]: That was done just in the sake of [possible] BC.
Fixed.

**Pull request merged**